File: internal/games/mahjong/session/state.go
package session

import (
	"errors"
	"time"

	"xgames/internal/games/mahjong/msg"
	"xgames/internal/games/mahjong/rule"
	"xgames/internal/platform/protocol"
)

// 机器人/挂机的出牌兜底超时（需略长于控制器主动出牌延迟 2~2.5s，控制器胜利则正常流转，否则由本兜底代打）
const afkActDelay = 3 * time.Second

func (gs *Session) handleAfk(playerID string, afk bool) error {
	if gs.phase == PhaseInit || gs.phase == PhaseEnded {
		return ErrGameNotStart
	}
	idx := gs.playerIdx(playerID)
	if idx == -1 {
		return ErrNotYourTurn
	}
	player := gs.players[idx]
	if player.IsBot || player.Afk == afk {
		return nil
	}

	player.Afk = afk
	gs.broadcastAfk(player, afk)
	gs.logger.Info("挂机状态变更", "room", gs.roomCode, "player", player.Name, "afk", afk)

	// 关闭托管：取消其挂起中的自动动作，局面转回真人交互（见 docs/architecture.md §10.3.3）
	gs.cancelPendingAutoFor(idx)

	if afk && gs.waitingForClaim && gs.claimPlayer == idx {
		// 托管开启时若正等待该玩家碰/杠/胡决策，立即按机器人策略代打
		gs.resolvePendingClaimForAfk(idx)
	}

	if gs.isCurrentTurnPlayer(idx) && gs.timer != nil {
		if afk {
			gs.startTurnTimer(afkActDelay)
		} else {
			gs.startTurnTimer(gs.timeouts.Turn)
		}
	}
	return nil
}

// resolvePendingClaimForAfk 托管开启时自动处理等待中的碰/杠/胡询问（机器人策略代打）
func (gs *Session) resolvePendingClaimForAfk(idx int) {
	gs.claimSeq++ // 使挂起的询问定时器失效

	if gs.claimIsSelfDraw {
		// 自摸胡：托管自动胡
		gs.waitingForClaim = false
		gs.claimIsSelfDraw = false
		gs.executeWin(idx, true)
		return
	}

	gs.waitingForClaim = false
	tile := gs.lastDiscard
	fromIdx := gs.lastDiscardFrom

	if rule.CanWinFromDiscardWithMelds(gs.hands[idx], tile, len(gs.melds[idx])) {
		gs.executeWin(idx, false)
		return
	}
	if rule.CanKongFromDiscard(gs.hands[idx], tile) && gs.botDecideKong(idx, tile) {
		gs.executeKong(idx, tile, fromIdx, false)
		return
	}
	if rule.CanPong(gs.hands[idx], tile) && gs.botDecidePong(idx, tile) {
		gs.executePong(idx, tile, fromIdx)
		return
	}

	// 无可执行动作，继续询问后续玩家
	gs.resolveClaimsAfter(fromIdx, idx)
}

func (gs *Session) broadcastAfk(player *Player, afk bool) {
	gs.broadcast(msg.MsgMjAfkChanged, msg.MjAfkChangedPayload{
		PlayerID: player.ID,
		Afk:      afk,
	})
}

func (gs *Session) enterAfkOnTimeout(idx int) {
	p := gs.players[idx]
	if p.IsBot || p.Afk || p.IsOffline {
		return
	}
	p.Afk = true
	gs.logger.Info("回合超时，自动转为挂机", "room", gs.roomCode, "player", p.Name)
	gs.broadcastAfk(p, true)
}

func (gs *Session) exitAfkOnAction(playerID string) {
	idx := gs.playerIdx(playerID)
	if idx == -1 {
		return
	}
	p := gs.players[idx]
	if p.IsBot || !p.Afk {
		return
	}
	p.Afk = false
	gs.logger.Info("主动行动，取消挂机", "room", gs.roomCode, "player", p.Name)
	gs.broadcastAfk(p, false)
	// 主动行动即回手动：取消其挂起中的自动动作（本次动作随后在对应 handler 中正常处理）
	gs.cancelPendingAutoFor(idx)
}

// --- 离线 ---

func (gs *Session) handleOffline(playerID string) {
	idx := gs.playerIdx(playerID)
	if idx == -1 {
		return
	}
	gs.players[idx].IsOffline = true

	if !gs.isCurrentTurnPlayer(idx) || gs.timer == nil {
		return
	}

	gs.pausedRest = gs.timeLeft()
	stopTimer(gs.timer)
	gs.timer = nil
	gs.turnPaused = true

	stopTimer(gs.offlineTimer)
	gs.offlineIdx = idx
	gs.offlineTimer = waitAfter(gs.timeouts.OfflineWait)

	gs.logger.Info("玩家离线，暂停计时", "room", gs.roomCode,
		"player", gs.players[idx].Name, "remaining", gs.pausedRest)
}

func (gs *Session) handleOnline(playerID string) {
	idx := gs.playerIdx(playerID)
	if idx == -1 {
		return
	}
	gs.players[idx].IsOffline = false

	if p := gs.players[idx]; !p.IsBot && p.Afk {
		p.Afk = false
		gs.broadcastAfk(p, false)
	}
	// 重连自动解除托管（绕过 handleAfk）：同样取消挂起中的自动动作，转回真人交互
	gs.cancelPendingAutoFor(idx)

	if gs.offlineTimer != nil && gs.offlineIdx == idx {
		stopTimer(gs.offlineTimer)
		gs.offlineTimer = nil
	}

	if !gs.turnPaused || !gs.isCurrentTurnPlayer(idx) {
		return
	}

	rest := gs.pausedRest
	gs.turnPaused = false
	gs.pausedRest = 0
	if rest <= 0 {
		gs.handleTurnTimeout()
		return
	}

	gs.timerStartedAt = timerNow()
	gs.turnTotal = rest
	stopTimer(gs.timer)
	gs.timer = waitAfter(rest)
	gs.logger.Info("玩家重连，恢复计时", "room", gs.roomCode,
		"player", gs.players[idx].Name, "remaining", rest)
}

func (gs *Session) handleOfflineTimeout() {
	idx := gs.offlineIdx
	if idx < 0 || idx >= len(gs.players) {
		return
	}
	player := gs.players[idx]
	if !player.IsOffline || !gs.isCurrentTurnPlayer(idx) {
		return
	}

	gs.logger.Info("玩家离线超时，自动托管", "room", gs.roomCode, "player", player.Name)

	if !player.Afk {
		player.Afk = true
		gs.broadcastAfk(player, true)
	}

	gs.autoplayFor(idx)
}

func (gs *Session) isCurrentTurnPlayer(idx int) bool {
	return gs.phase == PhasePlaying && gs.currentTurn == idx
}

// --- 状态 DTO ---

func (gs *Session) gameStateFor(playerID string) *msg.MjGameStateDTO {
	phase := "playing"
	if gs.phase == PhaseEnded {
		phase = "ended"
	}

	myIdx := gs.playerIdx(playerID)

	players := make([]msg.MjPlayerHandDTO, len(gs.players))
	for i, p := range gs.players {
		dto := msg.MjPlayerHandDTO{
			PlayerID:   p.ID,
			PlayerName: p.Name,
			Seat:       p.Seat,
			IsBot:      p.IsBot,
			Online:     !p.IsOffline,
			Afk:        p.Afk,
			HandCount:  len(gs.hands[i]),
			IsDealer:   i == gs.dealer,
			Score:      gs.scoreOf(i),
		}
		if i == myIdx {
			hand := make([]int, len(gs.hands[i]))
			copy(hand, gs.hands[i])
			sortHand(hand)
			dto.Hand = hand
		}
		for _, m := range gs.melds[i] {
			dto.Melds = append(dto.Melds, msg.MjMeldDTO{
				Type: int(m.Type),
				Tile: m.Tile,
				From: m.From,
			})
		}
		players[i] = dto
	}

	currentTurn := ""
	if gs.phase == PhasePlaying {
		currentTurn = gs.players[gs.currentTurn].ID
	}

	pools := make([][]int, len(gs.players))
	for i, pool := range gs.discardPools {
		pools[i] = append([]int(nil), pool...)
	}

	return &msg.MjGameStateDTO{
		Phase:        phase,
		Players:      players,
		CurrentTurn:  currentTurn,
		MyIndex:      myIdx,
		Dealer:       gs.dealer,
		WallRemain:   len(gs.wall),
		MeldNumber:   gs.meldNumber,
		DiscardPools: pools,
	}
}

// scoreOf 返回指定座位玩家的平台积分（未注入时为 0）
func (gs *Session) scoreOf(idx int) int {
	if idx < 0 || idx >= len(gs.scores) {
		return 0
	}
	return gs.scores[idx]
}

// --- 快照 ---

type Snapshot struct {
	RoomCode     string `json:"room_code"`
	Phase        string `json:"phase"`
	Players      []SnapshotPlayer `json:"players"`
	Wall         []int  `json:"wall"`
	Dealer       int    `json:"dealer"`
	CurrentTurn  int    `json:"current_turn"`
	MeldNumber   int    `json:"meld_number"`
	Hands        [][]int `json:"hands"`
	Melds        [][]rule.Meld `json:"melds"`
	DiscardPools [][]int `json:"discard_pools"`
	TimerRemaining int64 `json:"timer_remaining_ms"`
	TurnPaused   bool   `json:"turn_paused"`
	Scores       []int  `json:"scores,omitempty"` // 玩家平台积分（与 Players 索引对齐）
	TenpaiRounds []int  `json:"tenpai_rounds,omitempty"` // 每人听牌持续轮数（与 Players 索引对齐）
}

type SnapshotPlayer struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Seat      int    `json:"seat"`
	IsBot     bool   `json:"is_bot"`
	IsOffline bool   `json:"is_offline"`
	Afk       bool   `json:"afk,omitempty"`
}

func (gs *Session) snapshot() *Snapshot {
	phase := "playing"
	if gs.phase != PhasePlaying {
		phase = ""
	}

	players := make([]SnapshotPlayer, len(gs.players))
	for i, p := range gs.players {
		players[i] = SnapshotPlayer{
			ID: p.ID, Name: p.Name, Seat: p.Seat,
			IsBot: p.IsBot, IsOffline: p.IsOffline, Afk: p.Afk,
		}
	}

	var remaining int64
	switch {
	case gs.turnPaused:
		remaining = gs.pausedRest.Milliseconds()
	case gs.timer != nil:
		remaining = gs.timeLeft().Milliseconds()
	}

	hands := make([][]int, len(gs.hands))
	for i, h := range gs.hands {
		hands[i] = append([]int(nil), h...)
	}
	melds := make([][]rule.Meld, len(gs.melds))
	for i, m := range gs.melds {
		melds[i] = append([]rule.Meld(nil), m...)
	}
	pools := make([][]int, len(gs.discardPools))
	for i, p := range gs.discardPools {
		pools[i] = append([]int(nil), p...)
	}

	return &Snapshot{
		RoomCode:       gs.roomCode,
		Phase:          phase,
		Players:        players,
		Wall:           append([]int(nil), gs.wall...),
		Dealer:         gs.dealer,
		CurrentTurn:    gs.currentTurn,
		MeldNumber:     gs.meldNumber,
		Hands:          hands,
		Melds:          melds,
		DiscardPools:   pools,
		TimerRemaining: remaining,
		TurnPaused:     gs.turnPaused,
		Scores:         append([]int(nil), gs.scores...),
		TenpaiRounds:   append([]int(nil), gs.tenpaiRounds...),
	}
}

func RestoreFromSnapshot(snap *Snapshot, bc Broadcaster, sink ResultSink, to Timeouts) (*Session, error) {
	if snap == nil || snap.Phase == "" {
		return nil, errors.New("快照为空或不可恢复")
	}
	if len(snap.Players) < 2 {
		return nil, errors.New("快照玩家数不足")
	}

	players := make([]*Player, len(snap.Players))
	for i, sp := range snap.Players {
		players[i] = &Player{
			ID: sp.ID, Name: sp.Name, Seat: sp.Seat,
			IsBot: sp.IsBot, IsOffline: sp.IsOffline, Afk: sp.Afk,
		}
	}

	gs := New(snap.RoomCode, players, bc, sink, to)
	gs.restored = snap
	return gs, nil
}

func (gs *Session) applySnapshot(snap *Snapshot) {
	// 恢复局按恢复时刻重新计时，避免报告时长从零值时刻起算出现异常值
	gs.gameStartTime = time.Now()
	gs.phase = PhasePlaying
	gs.wall = append([]int(nil), snap.Wall...)
	gs.dealer = snap.Dealer
	gs.currentTurn = snap.CurrentTurn
	gs.meldNumber = snap.MeldNumber
	gs.hands = make([][]int, len(snap.Hands))
	for i, h := range snap.Hands {
		gs.hands[i] = append([]int(nil), h...)
	}
	gs.melds = make([][]rule.Meld, len(snap.Melds))
	for i, m := range snap.Melds {
		gs.melds[i] = append([]rule.Meld(nil), m...)
	}
	gs.discardPools = make([][]int, len(snap.DiscardPools))
	for i, p := range snap.DiscardPools {
		gs.discardPools[i] = append([]int(nil), p...)
	}
	if len(snap.TenpaiRounds) == len(gs.players) {
		gs.tenpaiRounds = append([]int(nil), snap.TenpaiRounds...)
	} else {
		gs.tenpaiRounds = make([]int, len(gs.players))
	}

	// 通知各玩家开局
	for i, p := range gs.players {
		gs.sendTo(p.ID, protocol.MsgGameStart, protocol.GameStartPayload{
			Players: gs.playerInfos(),
		})
		gs.sendHandTo(i)
	}

	gs.broadcast(msg.MsgMjTurn, msg.MjTurnPayload{
		PlayerID:   gs.players[gs.currentTurn].ID,
		Timeout:    int(gs.turnDelayFor(gs.currentTurn).Seconds()),
		MeldNumber: gs.meldNumber,
	})

	if snap.TimerRemaining > 0 && gs.timer != nil {
		rest := time.Duration(snap.TimerRemaining) * time.Millisecond
		gs.timerStartedAt = timerNow()
		gs.turnTotal = rest
		stopTimer(gs.timer)
		gs.timer = waitAfter(rest)
	}
}

File: web/src/games/mahjong/state.ts
import type { MjGameStateDTO, MjPlayerHandDTO } from '../../platform/protocol';

export const MJ_NUM_TYPES = 34; // 万(9) + 筒(9) + 条(9) + 字牌(7)
export const MJ_HAND_SIZE = 13;
export const MJ_HONOR_START = 27; // 字牌起始索引

export interface MjMeld {
  type: number; // 0=碰 1=杠 3=暗杠
  tile: number;
  from: number;
}

/** 面子有多少张牌 */
export function meldTileCount(type: number): number {
  return (type === 1 || type === 3) ? 4 : 3;
}

export interface MjPlayerState {
  playerId: string;
  playerName: string;
  seat: number;
  isBot: boolean;
  online: boolean;
  afk: boolean;
  handCount: number;
  hand: number[];
  melds: MjMeld[];
  isDealer: boolean;
  score: number;
}

export const mjState = {
  phase: '',
  players: [] as MjPlayerState[],
  currentTurn: '',
  myIndex: -1,
  dealer: 0,
  wallRemain: 0,
  meldNumber: 0,
  discardPools: [] as number[][],
  selectedTile: -1,
  actionAvail: false,
  canPong: false,
  canKong: false,
  canWin: false,
  actionTile: -1,
  selfDrew: false,

  reset() {
    this.phase = '';
    this.players = [];
    this.currentTurn = '';
    this.myIndex = -1;
    this.dealer = 0;
    this.wallRemain = 0;
    this.meldNumber = 0;
    this.discardPools = [];
    this.selectedTile = -1;
    this.actionAvail = false;
    this.canPong = false;
    this.canKong = false;
    this.canWin = false;
    this.actionTile = -1;
    this.selfDrew = false;
  },

  restore(dto: MjGameStateDTO) {
    this.phase = dto.phase;
    this.currentTurn = dto.current_turn;
    this.myIndex = dto.my_index;
    this.dealer = dto.dealer;
    this.wallRemain = dto.wall_remain;
    this.meldNumber = dto.meld_number;
    this.discardPools = dto.discard_pools || [];
    this.players = dto.players.map((p: MjPlayerHandDTO) => ({
      playerId: p.player_id,
      playerName: p.player_name,
      seat: p.seat,
      isBot: p.is_bot,
      online: p.online,
      afk: p.afk,
      handCount: p.hand_count,
      hand: p.hand || [],
      melds: (p.melds || []).map(m => ({ type: m.type, tile: m.tile, from: m.from })),
      isDealer: p.is_dealer,
      score: p.score ?? 0,
    }));
    this.selectedTile = -1;
    this.actionAvail = false;
    const me = this.myIndex >= 0 ? this.players[this.myIndex] : undefined;
    this.selfDrew = !!me && me.hand.length === MJ_HAND_SIZE + 1;
  },
};

/** 花色：0=万 1=筒 2=条 3=字牌 */
export function tileSuit(t: number): number {
  if (t >= MJ_HONOR_START) return 3;
  return Math.floor(t / 9);
}

/** 点数：数牌 1-9（返回 0-8），字牌 0-6 */
export function tileValue(t: number): number {
  if (t >= MJ_HONOR_START) return t - MJ_HONOR_START;
  return t % 9;
}

const HONOR_NAMES = ['东', '南', '西', '北', '中', '发', '白'];
const HONOR_SPEECH = ['东风', '南风', '西风', '北风', '红中', '发财', '白板'];
const SUIT_NAMES = ['万', '筒', '条'];

export function tileName(t: number): string {
  if (t >= MJ_HONOR_START) {
    return HONOR_NAMES[t - MJ_HONOR_START];
  }
  return `${tileValue(t) + 1}${SUIT_NAMES[tileSuit(t)]}`;
}

/** 语音播报用牌名：字牌用全称（东风/南风/西风/北风/红中/发财/白板） */
export function tileSpeech(t: number): string {
  if (t >= MJ_HONOR_START) {
    return HONOR_SPEECH[t - MJ_HONOR_START];
  }
  return tileName(t);
}

/** 牌面颜色：万红、筒蓝、条绿、字牌黑（中红、发绿、白灰） */
export function tileColor(t: number): number {
  const s = tileSuit(t);
  if (s === 0) return 0xe74c3c; // 万
  if (s === 1) return 0x3498db; // 筒
  if (s === 2) return 0x2ecc71; // 条
  // 字牌：东/南/西/北=0x2c3e50(深蓝), 中=0xe74c3c(红), 发=0x27ae60(绿), 白=0x7f8c8d(灰)
  const hv = t - MJ_HONOR_START;
  if (hv <= 3) return 0x2c3e50; // 东南西北
  if (hv === 4) return 0xe74c3c; // 中
  if (hv === 5) return 0x27ae60; // 发
  return 0x7f8c8d; // 白
}

/** 是否字牌 */
export function isHonor(t: number): boolean {
  return t >= MJ_HONOR_START;
}
